src/grep.js

```diff
@@ -62,23 +86,101 @@ function _grep(options, regex, files) {
       }
     } else {
       var lines = contents.split('\n');
+      var matches = [];
+
+      //  If before or after not given but context is, update values
+      if (options.context) {
+        if (options.beforeContext === false) {
+          options.beforeContext = options.context;
+        }
+        if (options.afterContext === false) {
+          options.afterContext = options.context;
+        }
+      }
+
       lines.forEach(function (line, index) {
         var matched = line.match(regex);
         if ((options.inverse && !matched) || (!options.inverse && matched)) {
-          var result = line;
-          if (options.lineNumber) {
-            result = '' + (index + 1) + ':' + line;
+          var lineNumber = index + 1;
+          var result = {};
+          if (matches.length > 0) {
+            // If the last result intersects, combine them
+            var last = matches[matches.length - 1];
+            var minimumLineNumber = Math.max(
+              1,
+              lineNumber - options.beforeContext - 1
+            );
+            if (
+              last.hasOwnProperty('' + lineNumber) ||
+              last.hasOwnProperty('' + minimumLineNumber)
+            ) {
+              result = last;
+            }
+          }
+          result[lineNumber] = {
+            line: line,
+            match: true,
+          };
+          if (options.beforeContext > 0) {
+            // Store the lines with their line numbers to check for overlap
+            lines
+              .slice(Math.max(index - options.beforeContext, 0), index)
+              .forEach(function (v, i, a) {
```

Comment:
It looks like this will involve a doubly-nested loop (O(n^2) complexity). Would it be possible to instead write this with a single loop (O(n) complexity)?

I think we can iterate through and maintain a queue of a max size B. Once we see a regex match, output the queue of size B (this is all of the "before" context) and output the current line. You can do a similar approach for the "after" context. If you see another regex match before the "after" context is complete, then just print the regex match and start over the with "after" queue.

Let me know if you feel up to trying this out, otherwise I can try to work on this after the PR lands.

Reply:
I’ve been trying to get it down to O(n) for a while today and still haven’t figured it out. Would you mind giving it a go?